### DISCUSSION TITLE
RegIf support for reading from/writing to Stream

### DISCUSSION THREAD

**Discussion by KireinaHoro** [SpinalHDL/SpinalHDL#1590] [Q&A]
The RegIf interface looks fairly feature complete, but one thing that's missing currently prevents me from switching to it (compared to my home-rolled register allocator): reading from / writing to `Stream[T]` and writing to `Flow[T]`.

Currently I'm using `BusSlaveFactory` with a very simple register allocator; most registers are normal control and status registers, but some have FIFO semantics.  One tricky bit of reading Stream with BusSlaveFactory is that, since many buses don't carry a "read strobe" signal, the entire bus width is always read.  This might trigger spurious FIFO pops when a register sharing the same bus width (64B for 512b AXI, for example) is read; therefore it requires special handling (i.e. putting FIFO registers in a 64B area where it is the only register).

It doesn't seem like there are any FIFO-style `AccessType` defined currently for `busif`.  Does this sound like a reasonable use case and are there plans to support this?  Thanks in advance!

**Comment by Readon** [comment]
Reading from Stream normally fail when valid is deasserted, many bus would not check if the data is really available. So that might require additional status signal to indicate it.

**Reply by KireinaHoro** [reply]
This is handled with readStreamNonBlocking in bus slave factory: it concatenates one bit (valid) to the payload.  The software is supposed to check the bit.

I have also a gadget that allows blocking read (up to a timeout) for buses that support it.

**Reply by Readon** [reply]
so in this cae, the data width should be less than bus data width?

**Reply by KireinaHoro** [reply]
I think `BusSlaveFactory` allows the payload to span multiple bus data widths:

https://github.com/SpinalHDL/SpinalHDL/blob/8ee0e9c942676fd740b22c888110f9fa696735da/lib/src/main/scala/spinal/lib/bus/misc/BusSlaveFactory.scala#L567-L592

Internally it uses `readMultiWord` anyways so the added bit is not an issue.

**Comment by Readon** [comment]
Also there is no backpressure of Flow also need a canDo signal.

**Reply by KireinaHoro** [reply]
Writing to a flow by definition should always succeed.  Reading is more problematic, but not something I need for now…
